Shopify/shopify-cli
# Consume bulk update in the theme assets API client

### WHY are these changes introduced?

We've introduced a new API that supports the upload of multiple assets in a single request. This PR relies on this new endpoint to optimize the `shopify theme push` and `shopify theme serve` commands.

### WHAT is this pull request doing?

This PR introduces an async API into the `ThemeAdminAPI`, so now it supports sync and async calls:
```ruby
# sync
status, body, response = api_client.put(path: path, body: body)

# async
api_client.put(path: path, body: body) do |status, resp_body, response|
  # ...
end
```

Also, it introduces the `ThemeAdminAPIThrottler` decorator, responsible for intercepting `PUT` requests in order to group them. The `ThemeAdminAPIThrottler` relies in the `ThemeAdminAPIThrottler::Bulk` class to perform requests in the new endpoint without blocking the `Syncer` (as some update operations are scheduled for a bulk requests, and can only be proceed when the bulk is ready).

Finally, after trying different tuninng approaches, I've decided to rely on:
- a pool of 20 threads
- a max bulk size of 20 assets
- a max bulk size of 10MB
- a timeout for new eligible PUT requests of 200ms

With those constraints, we've reached a `shopify theme push` command **6x faster** on our tests.

### How to test your changes?

- Create a theme with `shopify-dev theme init test_bulk`
- Run `shopify-dev theme push -u -t test_bulk_pr_2 --stable`
- Run `shopify-dev theme push -u -t test_bulk_pr_1`
- Notice is faster
- In a different directory, run `shopify-dev theme pull -t test_bulk_pr_1`
- In a different directory, run `shopify-dev theme pull -t test_bulk_pr_2`
- Compare both directories, they must be equal

### Post-release steps

None.

### Update checklist
